R/Quosure.R

```diff
@@ -0,0 +1,366 @@
+#' Reproducible class with environment and code.
+#'
+#' Reproducible class with environment and code.
+#' @name Quosure-class
+#' @rdname Quosure-class
+#' @slot code (`character`) to reproduce the environment
+#' @slot env (`environment`) environment which content was generated by the evaluation
+#'  of the `code` slot.
+#' @slot id (`integer`) random identifier of the code element to make sure uniqueness
+#'  when joining.
+#' @keywords internal
+setClass(
+  "Quosure",
+  slots = c(env = "environment", code = "character", id = "integer"),
+  prototype = list(env = new.env(parent = parent.env(.GlobalEnv)), code = character(0), id = integer(0))
+)
+
+#' It takes a `Quosure` class and returns TRUE if the input is valid
+#' @keywords internal
+setValidity("Quosure", function(object) {
+  if (length(object@code) != length(object@id)) {
+    "@code and @id slots must have the same length."
+  } else if (any(duplicated(object@id))) {
+    "@id contains duplicated values."
+  } else {
+    TRUE
+  }
+})
+
+#' Initialize `Quosure` object
+#'
+#' Initialize `Quosure` object with `code` and `env`. In order to have `Quosure` reproducible
+#' one needs to initialize with `env` which can be reproduced by the `code`. Alternatively, one
+#' can create an empty `Quosure` and evaluate the expressions in this object.
+#' @name new_quosure
+#'
+#' @param code (`character(1)` or `language`) code to evaluate. Accepts and stores comments also.
+#' @param env (`environment`) Environment being a result of the `code` evaluation.
```

Comment:
So it's not quite this - as there's nothing actually stopping anyone doing `q <- new_quosure(env = list2env(list(x = 5)), code = x <- 10)` it's more correctly the code which will be prepended to the `quosure::get_code` code and in order for reproducibility it should create the contents of `env`

(I wonder if we should have some tests showing this is ok - and also should we have an issue to have a check argument in the future so that teal.data could use quosures....)